[AIRFLOW-5911] Simplify lineage API and improve robustness

bolkedebruin opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5911
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This simplifies the lineage API which was needlessy cluttered.
You can now set "inlets='auto'" rather than "inlets={'auto': True}"
and airflow will figure out what to do.

Please note that this is not backward compatible. I'm assuming not many people (or none ;-) ) have been using the lineage api until now.
### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Unit tests updated. "sending" lineage data to the backend has been disabled at the moment and will be addressed in the follow up jira (see: https://issues.apache.org/jira/browse/AIRFLOW-5910)

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

Documentation has been updated.

cc @feng-tao

feng-tao commented on airflow/lineage/__init__.py:
```diff
@@ -18,18 +18,25 @@
+
 PIPELINE_OUTLETS = "pipeline_outlets"
 PIPELINE_INLETS = "pipeline_inlets"
+AUTO = "auto"
```
a comment for the global variable

bolkedebruin replied:
How'd you mean?

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -316,8 +316,8 @@ def __init__(
         do_xcom_push: bool = True,
-        inlets: Optional[Dict] = None,
-        outlets: Optional[Dict] = None,
+        inlets: Union[str, List[Any]] = [],
```
`List[str]` maybe?

ashb replied:
Oh no. I guess the "correct" type is

`Union[str, List[Union[str, airflow.lineage.entity.Dataset]]`?

bolkedebruin replied:
Good point. It should actually be Union[str, Dataset, BaseOperator]. But that will be a challenge (I cannot reference BaseOperator from within)

ashb replied:
You can use `Union[str, Dataset, "BaseOperator"]` I think.

bolkedebruin replied:
O didnt know that. Let's see. Was planning to have BaseOperator derive from Operator, but if yours works that is probably better

bolkedebruin replied:
Won't be able to do that. I'm having a circular dependency so I need to derive from Operator anyway.

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -412,25 +412,12 @@ def __init__(
-        if outlets:
-            self._outlets.update(outlets)
+        self._inlets = inlets
+        self._outlets = outlets
```
Why do we have `self._outlets` and `self.outlets`?

bolkedebruin replied:
Because self.[outlets,inlets] is a rendered version and self.[_outlets, _inlets] is the pre-rendered version

ashb commented on airflow/lineage/entity/dataset.py:
```diff
@@ -0,0 +1,25 @@
+
+@attr.s(auto_attribs=True)
+class Dataset:
+    pass
```
Given how small each of these classes are I'd be happy (and personally happier) to have them all live in airflow/lineage/entities.py

bolkedebruin replied:
Sure thing

ashb commented on docs/lineage.rst:
```diff
@@ -50,40 +51,43 @@ works.
 
-Outlets are defined as list of dataset ``{"datasets": [dataset1, dataset2]}``. Any fields for the dataset are templated with
-the context when the task is being executed.
+Outlets are defined as list of dataset ``[dataset1, dataset2]}``.
```
```suggestion
Outlets are defined as list of datasets ``[dataset1, dataset2]``.
```

ashb commented on docs/lineage.rst:
```diff
@@ -50,40 +51,43 @@ works.
+- can be configured to look for outlets from upstream tasks
 
-- can be configured to pick up outlets from direct upstream tasks ``{"auto": True}``
+- can be configured to pick up outlets from direct upstream tasks 'AUTO'
```
The way this is written people would do `inlets="AUTO"` which won't be right.

```suggestion
- can be configured to pick up outlets from direct upstream tasks ``airflow.lineage.AUTO``
```

bolkedebruin replied:
Should be 'auto' & 'AUTO' & 'airflow.lineage.AUTO'.

ashb commented on docs/lineage.rst:
```diff
@@ -50,40 +51,43 @@ works.
 Inlets can be manually defined by the following options:
 
-- by a list of dataset ``{"datasets": [dataset1, dataset2]}``
+- by a list of dataset
```
```suggestion
- by a list of airflow.lineage.entities.Dataset or a subclass
```

ashb commented on tests/lineage/test_lineage.py:
```diff
@@ -85,26 +85,26 @@ def test_lineage(self, _get_backend):
-        self.assertEqual(send_mock.call_count, 1)
-        send_mock.reset_mock()
+        # self.assertEqual(send_mock.call_count, 1)
+        # send_mock.reset_mock()
```
?

bolkedebruin replied:
I have disabled sending to a backend at the moment. This will be fixed in a follow up. I first want to get it right IN Airflow then work on external connections (Apache Atlas, Amundsen etc)

ashb commented on airflow/lineage/__init__.py:
```diff
@@ -92,51 +115,43 @@ def prepare_lineage(func):
 
-        for dataset in chain(self.inlets, self.outlets):
-            dataset.set_context(context)
+        self.outlets = [_get_instance(_as_rendered_dict(i, context))
```
I wonder if using `context['task'].render_template` would be better -- that way all the template rendering goes via the same path.

bolkedebruin replied:
I'll have a look, but render_template fails on non serializable fields, I prefer not to fail (or have it configirable here) for lineage as it probably confuses people

feng-tao commented on airflow/lineage/__init__.py:
```diff
@@ -92,51 +116,46 @@ def prepare_lineage(func):
 
-        if self._inlets['datasets']:  # pylint:disable=protected-access
-            self.inlets.extend(self._inlets['datasets'])  # pylint:disable=protected-access
+            self.inlets.extend([_get_instance(_as_rendered_dict(i, context))
```
if we put antlar to parse sql, will it put in the render_template func?

bolkedebruin replied:
I'm not entirely sure if I understand your question, but the context is the same as for render_template so outputs are the same. The difference is that this function will not fail if it find a non serializable field.

feng-tao commented on airflow/lineage/entities.py:
```diff
@@ -0,0 +1,97 @@
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class Table(Dataset):
```
didn't get the full context, but do you know why we put entities definitions here? And how easy to extend for custom entities?

bolkedebruin replied:
Airflow supports a light weight lineage. Lineage is the information that links Dataset A to Dataset B. Sometimes is important to give some hints to the lineage system to know what is is dealing with ie. File(Dataset A) -> Table(Dataset B).

I on purpose made the basic lineage info resemble Amundsen's model very closely. It will allow databuilder to enrich it or you could implement a Amundsen proxy in Airflow that allows it to push the information as soon as it becomes available. A proxy would also allow you to enrich the information for usage *in* Airflow, so pick up extra information and take actions based on those within the operator or do branching on them based on type of example.

Yes it's very easy to extend. See the papermill_operator for a Notebook example.

feng-tao reviewed (commented):
sorry for the late review @bolkedebruin

kaxil commented on airflow/lineage/__init__.py:
```diff
@@ -16,53 +16,85 @@
+    """
+    Class for serialized entities.
+    """
+    type: str = attr.ib()
```
Didn't knew about `attr` and `cattr` library, looks powerful. Thanks for it

Should we rename `type` to `_type` or `metadata_type` so that we don't use a built-in keyword

kaxil commented on airflow/lineage/__init__.py:
```diff
@@ -92,52 +120,49 @@ def prepare_lineage(func):
+            self._inlets = [self._inlets, ]                                                 # pylint:disable=protected-access, line-too-long # noqa: E501
 
-        self.log.debug("inlets: %s, outlets: %s", self.inlets, self.outlets)
+        if self._inlets and isinstance(self._inlets, list):                                 # pylint:disable=protected-access, line-too-long # noqa: E501
```
Can we add a comment on what is happening in this if condition with all the `union` and `intersection` stuff.

We should also try and get rid of `line-too-long` if possible as they are causing noise between code (atleast in the Github diff :D )

kaxil commented on airflow/operators/papermill_operator.py:
```diff
@@ -16,18 +16,22 @@
-    attributes = ['location', 'parameters']
+@attr.s(auto_attribs=True)
+class NoteBook(File):
+    type_hint: Optional[str] = "jupyter_notebook"
```
Let's add description and docstring for this class in the subsequent PRs

bolkedebruin replied:
Sure

kaxil commented on docs/lineage.rst:
```diff
@@ -50,40 +51,43 @@ works.
 Inlets can be manually defined by the following options:
 
-- by a list of dataset ``{"datasets": [dataset1, dataset2]}``
+- by a list of airflow.lineage.entities.Dataset or a subclass
```
```suggestion
- by a list of ``airflow.lineage.entities.Dataset`` or a subclass
```

kaxil commented on docs/lineage.rst:
```diff
@@ -50,40 +51,43 @@ works.
 .. note:: Operators can add inlets and outlets automatically if the operator supports it.
 
-In the example DAG task ``run_me_first`` is a BashOperator that takes 3 inlets: ``CAT1``, ``CAT2``, ``CAT3``, that are
+In the example DAG task ``run_this``(task_id ``run_me_first``) is a BashOperator that takes 3 inlets: ``CAT1``, ``CAT2``, ``CAT3``, that are
```
```suggestion
In the example DAG task ``run_this``(task_id=``run_me_first``) is a BashOperator that takes 3 inlets: ``CAT1``, ``CAT2``, ``CAT3``, that are
```

kaxil commented on setup.py:
```diff
@@ -359,7 +359,9 @@ def do_setup():
         install_requires=[
             'alembic>=1.0, <2.0',
             'argcomplete~=1.10',
+            'attrs==19.3.0',
```
```suggestion
            'attrs~=19.3.',
```

kaxil commented on setup.py:
```diff
@@ -359,7 +359,9 @@ def do_setup():
             'argcomplete~=1.10',
+            'attrs==19.3.0',
             'cached_property~=1.5',
+            'cattrs==0.9.0',
```
```suggestion
            'cattrs~=0.9',
```

feng-tao commented on airflow/lineage/backend/atlas/__init__.py:
```diff
@@ -1,102 +0,0 @@
-_host = conf.get("atlas", "host")
-
-
-class AtlasBackend(LineageBackend):  # pylint:disable=missing-docstring
```
where is the atlas backend now? Or is the lineage now just stored inside airflow metadata?

bolkedebruin replied:
The backend(s) will return but I want it to work cleanly in Airflow first. In any case I will expose the lineage information through a rest endpoint which you can use e.g. Admundsen to pick up the lineage by pulling. 

The backends will work two fold: 1) do push based lineage 2) allow for enrichment of the information available in aiflow. Think of In(Table("mytable", external=True)) that then resolves columns autmatically or other kind of information.

feng-tao commented on airflow/lineage/__init__.py:
```diff
@@ -92,52 +120,53 @@ def prepare_lineage(func):
+            self.inlets.extend(_inlets)
+
+        elif self._inlets:
+            raise AttributeError("inlets is not a list, operator, string or attr annotated object")
```
not sure if you have mentioned before, but for hive operator, we would like the outlet to be the s3 file which most of the time embedded in the hql. Is there a way to figure out the lineage?

bolkedebruin replied:
There are 3 ways to do this (some of it is future)

1. Set outlets to "outlets=File("s3a://XXX", key="my_key")" and do something like this in your hql spec  (pseudo)

```
SELECT * FROM {{ inlets['my_table'] }} INTO {{ outlets['my_key'] }}
```

Of course we might be able to expose the keys directly if they don't collide with existing keywords.

2. just set the inlets/outlets and make sure they match the HQL
3. We do antlr parsing and have the parser figure it out

2 and 3 are the least robust and hard to test.

feng-tao reviewed (commented):
got a bit more time to read though the pr now. Put a few more comments. Also I am super clear on where the lineage located? Previously you need to specify the backend(e.g atlas), where do you store it now? Is it generic in the airflow db?

feng-tao commented on airflow/lineage/__init__.py:
```diff
@@ -92,52 +120,53 @@ def prepare_lineage(func):
+
+            # pick up unique direct upstream task_ids if AUTO is specified
+            if AUTO.upper() in self._inlets or AUTO.lower() in self._inlets:
+                task_ids = task_ids.union(task_ids.symmetric_difference(self.upstream_task_ids))
```
didn't know symmetric_difference before :(

bolkedebruin replied:
With sets you can go crazy ;-)

feng-tao reviewed (approved):
lgtm , thanks @bolkedebruin

bolkedebruin committed:
[AIRFLOW-5911] Simplify lineage API and improve robustness

This simplifies the lineage API which was needlessy cluttered.
You can now set "inlets='auto'" rather than "inlets={'auto': True}"
and airflow will figure out what to do.

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>
Co-Authored-By: Kaxil Naik <kaxilnaik@gmail.com>
